codegen/src/main/scala/bsp/codegen/docs/MarkdownRenderer.scala

```diff
@@ -25,69 +32,88 @@ class MarkdownRenderer private (tree: DocTree, visited: MSet[ShapeId]) {
 
   def render: Lines = {
     val commonShapes =
-      if (tree.commonShapes.isEmpty) empty
+      if (tree.commonShapeIds.isEmpty) empty
       else
         lines(
           "## Common shapes",
           newline,
-          tree.commonShapes.foldMap(renderNode),
+          tree.commonShapeIds.foldMap(renderStructureNode),
           newline
         )
 
-    val services = tree.services.foldMap(renderNode)
-    commonShapes ++ services
+    // Server comes first because it's bigger.
+    val services = tree.services.sortBy(-_.operations.size).foldMap(renderServiceNode)
+    lines(
+      commonShapes,
+      services
+    )
   }
 
-  def renderNode(id: ShapeId): Lines = {
+  def renderStructureNode(id: ShapeId): Lines = {
     // We don't want to generate definitions for built-in primitives.
     if (id.getNamespace == "smithy.api") {
       return empty
     }
 
-    tree.docNodes(id) match {
-      case OperationDocNode(operation, inputNode, outputNode) =>
+    val node = tree.structures(id)
+
+    if (visited.contains(id)) empty
+    else {
+      visited.add(id)
+      lines(
+        renderStructureNodeDef(node.definition),
+        node.definition.members
+          .filter(_.getNamespace == id.getNamespace)
+          .foldMap(renderStructureNode)
+      )
+    }
+  }
+
+  def renderStructureNodeDef(definition: Def): Lines = {
+    lines(
+      s"#### ${definition.shapeId.getName()}",
+      newline,
+      documentation(definition.hints),
+      newline,
+      tsBlock(definition),
+      newline
+    )
+  }
+
+  def renderOperationNode(node: OperationDocNode): Lines = {
+    node match {
+      case OperationDocNode(operation, input, output) =>
```

Comment:
I suggest to avoid pattern matching on a single pattern.

```suggestion
    val OperationDocNode(operation, input, output) = node
```